File: 0542-01-matrix/0542-01-matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        rows, cols, q = len(mat), len(mat[0]), deque()
        dp = [[float("inf") for _ in range(0, cols)] for _ in range(0, rows)]
        for r in range(0, rows):
            for c in range(0, cols):
                if mat[r][c] == 0:
                    dp[r][c] = 0
                    q.append([r, c, 0])
        while q:
            r, c, d = q.popleft()
            for rr, cc in [[1, 0], [0, 1], [-1, 0], [0, -1]]:
                rrr = rr + r
                ccc = cc + c
                if rrr < 0 or rrr >= rows or ccc < 0 or ccc >= cols or 1 + d >= dp[rrr][ccc]:
                    continue
                dp[rrr][ccc] = 1 + d
                q.append([rrr, ccc, 1 + d])
        return dp

File: 0542-01-matrix/0542-01-matrix.js
/**
 * @param {number[][]} mat
 * @return {number[][]}
 */
var updateMatrix = function(mat) {
    const rows = mat.length
    const cols = mat[0].length
    const dp = Array(rows).fill(0).map(() => Array(cols).fill(Infinity))
    const q = []
    for (let r=0; r < rows; r++) {
        for (let c=0; c < cols; c++) {
            if (mat[r][c] === 0) {
                dp[r][c] = 0
                q.push([r, c, 0])
            }
        }
    }
    while (q.length) {
        const [r, c, d] = q.shift()
        for (const [rr, cc] of [[1, 0], [0, 1], [-1, 0], [0, -1]]) {
            const rrr = rr + r
            const ccc = cc + c
            if (!(rrr in mat) || !(ccc in mat[rrr]) || 1 + d >= dp[rrr][ccc]) continue
            dp[rrr][ccc] = 1 + d
            q.push([rrr, ccc, 1 + d])
        }
    }
    return dp
};